### PR TITLE
feat: Support knowhere tracing using OpenTelemetry

### PR DESCRIPTION
**Pull request by cydrain**
Issue: https://github.com/milvus-io/milvus/issues/21508

### CODE REVIEW

**Comment by cydrain** [comment]
rerun ut

**Comment by cydrain** [comment]
rerun ut

File: internal/querynodev2/segments/plan.go [hunk 1]
```diff
@@ -47,7 +49,18 @@ func createSearchPlanByExpr(ctx context.Context, col *Collection, expr []byte) (
 		return nil, errors.New("nil collection ptr, collectionID = " + fmt.Sprintln(col.id))
 	}
 	var cPlan C.CSearchPlan
-	status := C.CreateSearchPlanByExpr(col.collectionPtr, unsafe.Pointer(&expr[0]), (C.int64_t)(len(expr)), &cPlan)
+
+	span := trace.SpanFromContext(ctx)
```
Comment (zhengbuqian) [inline internal/querynodev2/segments/plan.go:53]: I think we need to end the span.

**Comment by cydrain** [comment]
rerun ut

File: internal/core/src/query/Plan.cpp [hunk 2]
```diff
@@ -69,17 +69,19 @@ ParsePlaceholderGroup(const Plan* plan,
 }
 
 std::unique_ptr<Plan>
-CreateSearchPlanByExpr(const Schema& schema,
+CreateSearchPlanByExpr(const tracer::TraceContext trace_ctx,
```
[the remark this answers is not in the document]
Comment (cydrain) [reply internal/core/src/query/Plan.cpp:72]: updated

File: internal/core/src/common/Tracer.cpp [hunk 3]
```diff
@@ -41,20 +43,20 @@ static std::shared_ptr<trace::TracerProvider> noop_trace_provider =
     std::make_shared<opentelemetry::trace::NoopTracerProvider>();
 
 void
-initTelementry(TraceConfig* config) {
+initTelemetry(const TraceConfig& cfg) {
     std::unique_ptr<opentelemetry::sdk::trace::SpanExporter> exporter;
```
[the remark this answers is not in the document]
Comment (cydrain) [reply internal/core/src/common/Tracer.cpp:47]: plan to remove exporter in a separate PR

File: internal/core/src/query/Plan.cpp [hunk 4]
```diff
@@ -69,17 +69,19 @@ ParsePlaceholderGroup(const Plan* plan,
...
     // Note: serialized_expr_plan is of binary format
     proto::plan::PlanNode plan_node;
     plan_node.ParseFromArray(serialized_expr_plan, size);
-    return ProtoParser(schema).CreatePlan(plan_node);
+    return ProtoParser(schema).CreatePlan(trace_ctx, plan_node);
 }
 
 std::unique_ptr<RetrievePlan>
-CreateRetrievePlanByExpr(const Schema& schema,
+CreateRetrievePlanByExpr(const tracer::TraceContext trace_ctx,
```
[the remark this answers is not in the document]
Comment (cydrain) [reply internal/core/src/query/Plan.cpp:83]: updated

File: internal/core/src/common/Tracer.cpp [hunk 5]
```diff
@@ -118,20 +120,52 @@ CloseRootSpan() {
...
+}
+
+bool
+EmptySpanID(const TraceContext* ctx) {
+    return isEmptyID(ctx->spanID, trace::SpanId::kSize);
+}
+
+std::string
+GetTraceIDAsStr(const TraceContext* ctx) {
+    if (ctx != nullptr && ctx->traceID != nullptr) {
```
[the remark this answers is not in the document]
Comment (cydrain) [reply internal/core/src/common/Tracer.cpp:153]: updated

File: internal/core/src/query/Plan.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (cydrain) [reply internal/core/src/query/Plan.cpp:83]: updated

File: internal/core/src/common/Tracer.cpp [hunk 6]
```diff
@@ -81,22 +83,22 @@ initTelementry(TraceConfig* config) {
...
 GetTracer() {
     auto provider = trace::Provider::GetTracerProvider();
-    return provider->GetTracer("segcore", OPENTELEMETRY_SDK_VERSION);
+    return provider->GetTracer(TRACE_SERVICE_SEGCORE,
+                               OPENTELEMETRY_SDK_VERSION);
 }
 
 std::shared_ptr<trace::Span>
-StartSpan(const std::string& name, TraceContext* parentCtx) {
+StartSpan(const std::string& name, TraceContext* ctx) {
```
[the remark this answers is not in the document]
Comment (cydrain) [reply internal/core/src/common/Tracer.cpp:91]: updated

File: internal/core/src/index/VectorDiskIndex.cpp [hunk 7]
```diff
@@ -325,6 +325,17 @@ VectorDiskAnnIndex<T>::Query(const DatasetPtr dataset,
 
     knowhere::Json search_config = search_info.search_params_;
 
+    // save trace context into search conf
+    if (search_info.trace_ctx_.traceID != nullptr &&
```
[the remark this answers is not in the document]
Comment (cydrain) [reply internal/core/src/index/VectorDiskIndex.cpp]: updated

**Review by congqixia — COMMENTED** [review]
why shall we put trace ctx in search info? It looks weird since tracing is not related to search/query only

**Comment by cydrain** [comment]
> why shall we put trace ctx in search info? It looks weird since tracing is not related to search/query only

@congqixia 
This doc explains why I add trace_ctx into search_info
https://zilliverse.feishu.cn/docx/X4zodgAoFoJ0I9xIuDzcwgzFnYf?from=from_copylink

**Pull request closed without merging**